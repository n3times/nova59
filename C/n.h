#include <stdbool.h>

#define ABS(x)          ( (x) < 0 ? (-x) : (x) )
#define MIN(x, y)       ( (x) < (y) ? (x) : (y) )
#define MAX(x, y)       ( (x) > (y) ? (x) : (y) )
#define N_ELEMS(array)  ( sizeof(array) / sizeof((array)[0]) )

typedef struct n_s {
  long long mant;
  int exp;
} n_t;

typedef enum mode_e {
  FLOAT,
  EE,
  ENG
} mode_t;

static n_t N_0   = {             0LL,   0 };
static n_t N_INF = { 9999999999999LL,  99 };
static n_t N_EPS = { 1000000000000LL, -99 };

// n number
// fix in 0..9
// mode in { FLOAT, EE, ENG)
// str_out must have at least 16 characters
// 
// Returns representation of number as string:
// - optional "-"
// - followed by float, with exactly 1 "."
// - optionally followed by exponent: " " or "-", and 2 digits
// - optionally followed by overflow: " ?".
void n2s(n_t n, int fix, mode_t mode, char *str_out);

// s is a string:
// - optional "-"
// - followed by float, with at most 1 "."
// - optionally followed by exponent: " " or "-", and 2 digits
// - optionally followed by overflow: " ?".
n_t s2n(char *s, bool *err);
