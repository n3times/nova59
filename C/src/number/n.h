#ifndef N_H
#define N_H

#include <stdbool.h>


/******************************************************************************
 *
 *  TYPES.
 *
 ******************************************************************************/

/**
 * TI-59 number.
 *
 * For non zero numbers:
 * - mantissa is a positive or negative integer, with exactly 13 digits.
 * - exponent is in -99..99.
 *
 * 0 is represented as { 0, 0 }.
 */
typedef struct n_s {
  long long mant;
  int exp;
} n_t;

/** Number notation. */
typedef enum notation_e {
  FLOAT,  // Float if possible, scientific for small and big numbers.
  SCI,    // Scientific notation.
  ENG     // Engineering notation.
} notation_t;

/** Trigonometric mode. */
typedef enum trig_e {
  RAD,  // Radians.
  DEG,  // Degrees.
  GRAD  // Gradians.
} trig_t;


/******************************************************************************
 *
 *  CONSTANTS.
 *
 ******************************************************************************/

extern n_t N_0;    // 0.
extern n_t N_1;    // 1.
extern n_t N_PI;   // Pi.
extern n_t N_INF;  // Infinity, largest number.
extern n_t N_EPS;  // Epsilon, smallest positive number.


/******************************************************************************
 *
 *  NOTE ON ERRORS.
 *
 *  For functions, if "err" parameter is non null, it will be set to true if an
 *  error occurs, and false otherwise.
 *  An error occurs when some parameter is not in the function domain ("square
 *  root of -1") or because of underflow or overflow ("square of 10^80").
 *
 ******************************************************************************/


/******************************************************************************
 *
 *  ARITHMETIC OPERATORS.
 *
 ******************************************************************************/

/** Addition. */
n_t n_plus(n_t n1, n_t n2, bool *err);

/** Substraction. */
n_t n_minus(n_t n1, n_t n2, bool *err);

/** Multiplication. */
n_t n_times(n_t n1, n_t n2, bool *err);

/** Division. */
n_t n_div(n_t n1, n_t n2, bool *err);

/** Exponentiation. */
n_t n_pow(n_t n1, n_t n2, bool *err);

/** Root extraction. */
n_t n_ipow(n_t n1, n_t n2, bool *err);


/******************************************************************************
 *
 *  MATH FUNCTIONS.
 *
 ******************************************************************************/

/** -n. */
n_t n_chs(n_t n);

/** |n|. */
n_t n_abs(n_t n);

/** 0, 1 or -1, depending on whether n == 0, n > 0 or n < 0. */
n_t n_sign(n_t n);

/** The integral part of n: n_int(-3.6) == -3, n_int(3.6) == 3. */
n_t n_int(n_t n);

/** The fractional part of n: n_frac(-3.6) == -0.6, n_frac(3.6) == 0.6. */
n_t n_frac(n_t n);

/** n * n. */
n_t n_square(n_t n, bool *err);

/** 1/n. */
n_t n_1_x(n_t n, bool *err);

/** Square root of n. */
n_t n_sqrt(n_t n, bool *err);

/** Natural log. */
n_t n_ln(n_t n, bool *err);

/** Log in base 10. */
n_t n_log(n_t n, bool *err);

/** e^n. */
n_t n_exp(n_t n, bool *err);

/** 10^n. */
n_t n_pow10(n_t n, bool *err);


/******************************************************************************
 *
 *  TRIGONOMETRIC FUNCTIONS.
 *
 ******************************************************************************/

/** sine. */
n_t n_sin(n_t n, trig_t mode, bool *err);

/** cosine. */
n_t n_cos(n_t n, trig_t mode, bool *err);

/** tangent. */
n_t n_tan(n_t n, trig_t mode, bool *err);

/** arcsine. */
n_t n_asin(n_t n, trig_t mode, bool *err);

/** arccosine. */
n_t n_acos(n_t n, trig_t mode, bool *err);

/** arctangent. */
n_t n_atan(n_t n, trig_t mode, bool *err);


/******************************************************************************
 *
 *  SPECIAL FUNCTIONS.
 *
 ******************************************************************************/

/** Converts degrees/minutes/seconds to decimal degrees. */
n_t n_dms(n_t n, int fix, notation_t notation, bool *err);

/** Converts decimal degrees to degrees/minutes/seconds. */
n_t n_idms(n_t n, int fix, notation_t notation, bool *err);

/** Converts polar coordinates to rectangular coordinates. */
void n_p_r(n_t rho, n_t theta, n_t *x_out, n_t *y_out, trig_t mode, bool *err);

/** Converts rectangular coordinates to polar coordinates. */
void n_r_p(n_t x, n_t y, n_t *rho_out, n_t *theta_out, trig_t mode, bool *err);


/******************************************************************************
 *
 *  CONVERSION WITH STRINGS.
 *
 *  String representation of numbers:
 *  - optional "-"
 *  - followed by digits and at most 1 "."
 *  - optionally followed by exponent: " " or "-", and 2 digits
 *  - optionally followed by overflow: " ?".
 *
 ******************************************************************************/

/** Number to string. String 'str_out' must have at least 16 characters. */
void n2s(n_t n, int fix, notation_t notation, char *str_out);

/** String to number. Sets error if under/overflow. */
n_t s2n(char *s, bool *err);


/******************************************************************************
 *
 *  CONVERSION WITH DOUBLES.
 *
 ******************************************************************************/

/** Number to double. */
double n2d(n_t n);

/** Double to number. Sets error if under/overflow. */
n_t d2n(double d, bool *err);

#endif  // N_H
