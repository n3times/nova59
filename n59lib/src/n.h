#ifndef N_H
#define N_H

#include <stdbool.h>


/**
 * Library that defines TI-59 numbers as well as arithmetic operators and
 * mathematical functions on those numbers.
 *
 * Note on errors:
 *
 * For functions, if 'err_out' parameter is non null, it will be set to true
 * if an error occurs, and false otherwise.
 * An error occurs when some parameter is not in the function domain ("square
 * root of -1") or because of underflow or overflow ("square of 10^80").
 */


/******************************************************************************
 *
 *  TYPES.
 *
 ******************************************************************************/

/**
 * TI-59 number.
 *
 * For non zero numbers,
 * - the mantissa is a positive or negative integer, with exactly 13 digits
 * - the exponent is an integer in -99..99
 *
 * 0 is represented as { 0, 0 }.
 *
 * Note: Prefer n_make(mant, exp) to directly modifying mant and exp.
 */
typedef struct n_s {
  long long mant;
  int exp;
} n_t;

/** Display formats. */
typedef enum format_e {
  FLOAT,  // Float if possible, scientific for small and big numbers.
  SCI,    // Scientific notation.
  ENG     // Engineering notation.
} format_t;

/** Trigonometric modes. */
typedef enum trig_e {
  RAD,  // Radians.
  DEG,  // Degrees.
  GRAD  // Gradians.
} trig_t;


/******************************************************************************
 *
 *  CONSTANTS.
 *
 ******************************************************************************/

extern n_t N_0;    // 0.
extern n_t N_1;    // 1.
extern n_t N_PI;   // Pi.
extern n_t N_INF;  // Infinity, largest number.
extern n_t N_EPS;  // Epsilon, smallest positive number.


/******************************************************************************
 *
 *  UTILITY METHODS.
 *
 ******************************************************************************/

/**
 * Preferred method to make a TI-59 number.
 *
 * Asserts that mant and exp are in the correct range.
 */
n_t n_make(long long mant, int exp);

/** Returns true if n1 and n2 are equal. */
bool n_equals(n_t n1, n_t n2);

/** Returns true if n is zero. */
bool n_is_zero(n_t n);


/******************************************************************************
 *
 *  ARITHMETIC OPERATORS.
 *
 ******************************************************************************/

/** Addition. */
n_t n_plus(n_t n1, n_t n2, bool *err_out);

/** Substraction. */
n_t n_minus(n_t n1, n_t n2, bool *err_out);

/** Multiplication. */
n_t n_times(n_t n1, n_t n2, bool *err_out);

/** Division. */
n_t n_div(n_t n1, n_t n2, bool *err_out);

/** Exponentiation. */
n_t n_pow(n_t n1, n_t n2, bool *err_out);

/** Root extraction. */
n_t n_ipow(n_t n1, n_t n2, bool *err_out);


/******************************************************************************
 *
 *  MATH FUNCTIONS.
 *
 ******************************************************************************/

/** -n. */
n_t n_chs(n_t n);

/** |n|. */
n_t n_abs(n_t n);

/** 0, 1 or -1, depending on whether n == 0, n > 0 or n < 0. */
n_t n_sign(n_t n);

/** The integral part of n: n_int(-3.6) == -3, n_int(3.6) == 3. */
n_t n_int(n_t n);

/** The fractional part of n: n_frac(-3.6) == -0.6, n_frac(3.6) == 0.6. */
n_t n_frac(n_t n);

/** n * n. */
n_t n_square(n_t n, bool *err_out);

/** 1/n. */
n_t n_1_x(n_t n, bool *err_out);

/** Square root of n. */
n_t n_sqrt(n_t n, bool *err_out);

/** Natural log. */
n_t n_ln(n_t n, bool *err_out);

/** Log in base 10. */
n_t n_log(n_t n, bool *err_out);

/** e^n. */
n_t n_exp(n_t n, bool *err_out);

/** 10^n. */
n_t n_pow10(n_t n, bool *err_out);


/******************************************************************************
 *
 *  TRIGONOMETRIC FUNCTIONS.
 *
 ******************************************************************************/

/** sine. */
n_t n_sin(n_t n, trig_t mode, bool *err_out);

/** cosine. */
n_t n_cos(n_t n, trig_t mode, bool *err_out);

/** tangent. */
n_t n_tan(n_t n, trig_t mode, bool *err_out);

/** arcsine. Range is -90..90 (in DEG mode). */
n_t n_asin(n_t n, trig_t mode, bool *err_out);

/** arccossine. Range is 0..180 (in DEG mode). */
n_t n_acos(n_t n, trig_t mode, bool *err_out);

/** arctangent. Range is -90..90 (in DEG mode). */
n_t n_atan(n_t n, trig_t mode, bool *err_out);


/******************************************************************************
 *
 *  SPECIAL FUNCTIONS.
 *
 ******************************************************************************/

/**
 * Converts degrees/minutes/seconds to decimal degrees.
 *
 * Only the digits of n that are visible on the TI-59 display are considered
 * for the conversion. For example, if format is FLOAT and fix is 2, 3.1549 will
 * be first trimmed down to 3.15 and then converted to 3.25.
 */
n_t n_dms(n_t n, int fix, format_t format, bool *err_out);

/**
 * Converts decimal degrees to degrees/minutes/seconds.
 *
 * Only the digits of n that are visible on the TI-59 display are considered
 * for the conversion. For example, if format is FLOAT and fix is 2, 3.2549 will
 * be first trimmed down to 3.25 and then converted to 3.15.
 */
n_t n_idms(n_t n, int fix, format_t format, bool *err_out);

/** Converts polar coordinates to rectangular coordinates. */
void n_p_r(n_t rho, n_t theta, n_t *x_out, n_t *y_out, trig_t mode, bool *err_out);

/**
 * Converts rectangular coordinates to polar coordinates.
 *
 * theta_out is in range -90..270
 */
void n_r_p(n_t x, n_t y, n_t *rho_out, n_t *theta_out, trig_t mode, bool *err_out);


/******************************************************************************
 *
 *  CONVERSION TO/FROM STRINGS.
 *
 ******************************************************************************/

/**
 * Number to string.
 *
 * This is the number as it would appear on the display with leading and
 * trailing spaces trimmed:
 * - "-" if negative number
 * - followed by digits and exactly 1 "."
 *
 * For example: for pi, with fix 2, format SCI: "3.14 00".
 *
 * String 'str_out' must have at least 14 characters.
 * Sets error if overflow, that is when the display would blink on a TI-59.
 */
void n2s(n_t n, int fix, format_t format, char *str_out, bool *err_out);

/**
 * String to number.
 *
 * String should be composed of:
 * - a float number such as "1", "12.34", "-12.", "0.1" or "-.1"
 * - followed by an exponent such as "2", "04" or "-99"
 *
 * There may be spaces at the beginning and end of the string, and between the
 * float number and the exponent. If the the exponent is positive, a space is
 * required between the float and the exponent.
 *
 * Sets error if under/overflow or wrong formatting.
 */
n_t s2n(char *s, bool *err_out);


/******************************************************************************
 *
 *  CONVERSION TO/FROM DOUBLES.
 *
 ******************************************************************************/

/** Number to double. */
double n2d(n_t n);

/**
 * Double to number.
 *
 * Sets error if under/overflow.
 */
n_t d2n(double d, bool *err_out);

#endif  // N_H
