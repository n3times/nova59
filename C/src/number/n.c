#include "n.h"

n_t N_0   = {             0LL,   0 };
n_t N_1   = {             1LL,   0 };
n_t N_PI  = { 3141592653590LL,   0 };
n_t N_INF = { 9999999999999LL,  99 };
n_t N_EPS = { 1000000000000LL, -99 };
