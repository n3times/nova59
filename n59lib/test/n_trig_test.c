#include <stdio.h>

#include "n.h"

int main() {
  n_t n = { 9000000000000LL, 0 };
  printf("Forensics asin(acos(atan(tan(cos(sin(9)))))):\n%lld %d => ", n.mant, n.exp);
  n_t res = n = n_asin(n_acos(n_atan(n_tan(n_cos(n_sin(n, DEG,
      NULL), DEG, NULL), DEG, NULL), DEG, NULL), DEG, NULL), DEG, NULL);
  printf("%lld %d\n\n", res.mant, res.exp);
  trig_t mode = GRAD;
  for (double i = -800; i < 800; i += 50) {
    printf("%f sin %e\n", i, n2d(n_sin(d2n(i, NULL), mode, NULL)));
    printf("%f cos %e\n", i, n2d(n_cos(d2n(i, NULL), mode, NULL)));
    printf("%f tan %e\n", i, n2d(n_tan(d2n(i, NULL), mode, NULL)));
  }
  n_t n2 = { 3141592653590LL, 0 };
  mode = RAD;
  printf("%lld sin %e\n", n2.mant, n2d(n_sin(n2, mode, NULL)));
  printf("%lld cos %e\n", n2.mant, n2d(n_cos(n2, mode, NULL)));
  printf("%lld tan %e\n", n2.mant, n2d(n_tan(n2, mode, NULL)));
  n2.mant /= 2;
  printf("%lld sin %e\n", n2.mant, n2d(n_sin(n2, mode, NULL)));
  printf("%lld cos %e\n", n2.mant, n2d(n_cos(n2, mode, NULL)));
  printf("%lld tan %e\n", n2.mant, n2d(n_tan(n2, mode, NULL)));
}
