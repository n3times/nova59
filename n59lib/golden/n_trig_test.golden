Forensics asin(acos(atan(tan(cos(sin(9)))))):
9000000000000 0 => 8999999981719 0

Deg:  sin  -720.000000 =>  0000000000000 00
Deg:  sin  -675.000000 =>  7071067811866-01
Deg:  sin  -630.000000 =>  1000000000000 00
Deg:  sin  -585.000000 =>  7071067811864-01
Deg:  sin  -540.000000 =>  0000000000000 00
Deg:  sin  -495.000000 => -7071067811867-01
Deg:  sin  -450.000000 => -1000000000000 00
Deg:  sin  -405.000000 => -7071067811863-01
Deg:  sin  -360.000000 =>  0000000000000 00
Deg:  sin  -315.000000 =>  7071067811866-01
Deg:  sin  -270.000000 =>  1000000000000 00
Deg:  sin  -225.000000 =>  7071067811864-01
Deg:  sin  -180.000000 =>  0000000000000 00
Deg:  sin  -135.000000 => -7071067811867-01
Deg:  sin   -90.000000 => -1000000000000 00
Deg:  sin   -45.000000 => -7071067811863-01
Deg:  sin     0.000000 =>  0000000000000 00
Deg:  sin    45.000000 =>  7071067811866-01
Deg:  sin    90.000000 =>  1000000000000 00
Deg:  sin   135.000000 =>  7071067811864-01
Deg:  sin   180.000000 =>  0000000000000 00
Deg:  sin   225.000000 => -7071067811867-01
Deg:  sin   270.000000 => -1000000000000 00
Deg:  sin   315.000000 => -7071067811863-01
Deg:  sin   360.000000 =>  0000000000000 00
Deg:  sin   405.000000 =>  7071067811866-01
Deg:  sin   450.000000 =>  1000000000000 00
Deg:  sin   495.000000 =>  7071067811864-01
Deg:  sin   540.000000 =>  0000000000000 00
Deg:  sin   585.000000 => -7071067811867-01
Deg:  sin   630.000000 => -1000000000000 00
Deg:  sin   675.000000 => -7071067811863-01

Deg:  cos  -720.000000 =>  1000000000000 00
Deg:  cos  -675.000000 =>  7071067811865-01
Deg:  cos  -630.000000 =>  0000000000000 00
Deg:  cos  -585.000000 => -7071067811867-01
Deg:  cos  -540.000000 => -1000000000000 00
Deg:  cos  -495.000000 => -7071067811864-01
Deg:  cos  -450.000000 =>  0000000000000 00
Deg:  cos  -405.000000 =>  7071067811868-01
Deg:  cos  -360.000000 =>  1000000000000 00
Deg:  cos  -315.000000 =>  7071067811865-01
Deg:  cos  -270.000000 =>  0000000000000 00
Deg:  cos  -225.000000 => -7071067811867-01
Deg:  cos  -180.000000 => -1000000000000 00
Deg:  cos  -135.000000 => -7071067811864-01
Deg:  cos   -90.000000 =>  0000000000000 00
Deg:  cos   -45.000000 =>  7071067811868-01
Deg:  cos     0.000000 =>  1000000000000 00
Deg:  cos    45.000000 =>  7071067811865-01
Deg:  cos    90.000000 =>  0000000000000 00
Deg:  cos   135.000000 => -7071067811867-01
Deg:  cos   180.000000 => -1000000000000 00
Deg:  cos   225.000000 => -7071067811864-01
Deg:  cos   270.000000 =>  0000000000000 00
Deg:  cos   315.000000 =>  7071067811868-01
Deg:  cos   360.000000 =>  1000000000000 00
Deg:  cos   405.000000 =>  7071067811865-01
Deg:  cos   450.000000 =>  0000000000000 00
Deg:  cos   495.000000 => -7071067811867-01
Deg:  cos   540.000000 => -1000000000000 00
Deg:  cos   585.000000 => -7071067811864-01
Deg:  cos   630.000000 =>  0000000000000 00
Deg:  cos   675.000000 =>  7071067811868-01

Deg:  tan  -720.000000 =>  0000000000000 00
Deg:  tan  -675.000000 =>  1000000000000 00
Deg:  tan  -630.000000 =>  9999999999999 99
Deg:  tan  -585.000000 => -9999999999997-01
Deg:  tan  -540.000000 =>  0000000000000 00
Deg:  tan  -495.000000 =>  1000000000001 00
Deg:  tan  -450.000000 =>  9999999999999 99
Deg:  tan  -405.000000 => -9999999999993-01
Deg:  tan  -360.000000 =>  0000000000000 00
Deg:  tan  -315.000000 =>  1000000000000 00
Deg:  tan  -270.000000 =>  9999999999999 99
Deg:  tan  -225.000000 => -9999999999997-01
Deg:  tan  -180.000000 =>  0000000000000 00
Deg:  tan  -135.000000 =>  1000000000001 00
Deg:  tan   -90.000000 =>  9999999999999 99
Deg:  tan   -45.000000 => -9999999999993-01
Deg:  tan     0.000000 =>  0000000000000 00
Deg:  tan    45.000000 =>  1000000000000 00
Deg:  tan    90.000000 =>  9999999999999 99
Deg:  tan   135.000000 => -9999999999997-01
Deg:  tan   180.000000 =>  0000000000000 00
Deg:  tan   225.000000 =>  1000000000001 00
Deg:  tan   270.000000 =>  9999999999999 99
Deg:  tan   315.000000 => -9999999999993-01
Deg:  tan   360.000000 =>  0000000000000 00
Deg:  tan   405.000000 =>  1000000000000 00
Deg:  tan   450.000000 =>  9999999999999 99
Deg:  tan   495.000000 => -9999999999997-01
Deg:  tan   540.000000 =>  0000000000000 00
Deg:  tan   585.000000 =>  1000000000001 00
Deg:  tan   630.000000 =>  9999999999999 99
Deg:  tan   675.000000 => -9999999999993-01

Grad: sin  -800.000000 =>  0000000000000 00
Grad: sin  -750.000000 =>  7071067811866-01
Grad: sin  -700.000000 =>  1000000000000 00
Grad: sin  -650.000000 =>  7071067811864-01
Grad: sin  -600.000000 =>  0000000000000 00
Grad: sin  -550.000000 => -7071067811867-01
Grad: sin  -500.000000 => -1000000000000 00
Grad: sin  -450.000000 => -7071067811863-01
Grad: sin  -400.000000 =>  0000000000000 00
Grad: sin  -350.000000 =>  7071067811866-01
Grad: sin  -300.000000 =>  1000000000000 00
Grad: sin  -250.000000 =>  7071067811864-01
Grad: sin  -200.000000 =>  0000000000000 00
Grad: sin  -150.000000 => -7071067811867-01
Grad: sin  -100.000000 => -1000000000000 00
Grad: sin   -50.000000 => -7071067811863-01
Grad: sin     0.000000 =>  0000000000000 00
Grad: sin    50.000000 =>  7071067811866-01
Grad: sin   100.000000 =>  1000000000000 00
Grad: sin   150.000000 =>  7071067811864-01
Grad: sin   200.000000 =>  0000000000000 00
Grad: sin   250.000000 => -7071067811867-01
Grad: sin   300.000000 => -1000000000000 00
Grad: sin   350.000000 => -7071067811863-01
Grad: sin   400.000000 =>  0000000000000 00
Grad: sin   450.000000 =>  7071067811866-01
Grad: sin   500.000000 =>  1000000000000 00
Grad: sin   550.000000 =>  7071067811864-01
Grad: sin   600.000000 =>  0000000000000 00
Grad: sin   650.000000 => -7071067811867-01
Grad: sin   700.000000 => -1000000000000 00
Grad: sin   750.000000 => -7071067811863-01

Grad: cos  -800.000000 =>  1000000000000 00
Grad: cos  -750.000000 =>  7071067811865-01
Grad: cos  -700.000000 =>  0000000000000 00
Grad: cos  -650.000000 => -7071067811867-01
Grad: cos  -600.000000 => -1000000000000 00
Grad: cos  -550.000000 => -7071067811864-01
Grad: cos  -500.000000 =>  0000000000000 00
Grad: cos  -450.000000 =>  7071067811868-01
Grad: cos  -400.000000 =>  1000000000000 00
Grad: cos  -350.000000 =>  7071067811865-01
Grad: cos  -300.000000 =>  0000000000000 00
Grad: cos  -250.000000 => -7071067811867-01
Grad: cos  -200.000000 => -1000000000000 00
Grad: cos  -150.000000 => -7071067811864-01
Grad: cos  -100.000000 =>  0000000000000 00
Grad: cos   -50.000000 =>  7071067811868-01
Grad: cos     0.000000 =>  1000000000000 00
Grad: cos    50.000000 =>  7071067811865-01
Grad: cos   100.000000 =>  0000000000000 00
Grad: cos   150.000000 => -7071067811867-01
Grad: cos   200.000000 => -1000000000000 00
Grad: cos   250.000000 => -7071067811864-01
Grad: cos   300.000000 =>  0000000000000 00
Grad: cos   350.000000 =>  7071067811868-01
Grad: cos   400.000000 =>  1000000000000 00
Grad: cos   450.000000 =>  7071067811865-01
Grad: cos   500.000000 =>  0000000000000 00
Grad: cos   550.000000 => -7071067811867-01
Grad: cos   600.000000 => -1000000000000 00
Grad: cos   650.000000 => -7071067811864-01
Grad: cos   700.000000 =>  0000000000000 00
Grad: cos   750.000000 =>  7071067811868-01

Grad: tan  -800.000000 =>  0000000000000 00
Grad: tan  -750.000000 =>  1000000000000 00
Grad: tan  -700.000000 =>  9999999999999 99
Grad: tan  -650.000000 => -9999999999997-01
Grad: tan  -600.000000 =>  0000000000000 00
Grad: tan  -550.000000 =>  1000000000001 00
Grad: tan  -500.000000 =>  9999999999999 99
Grad: tan  -450.000000 => -9999999999993-01
Grad: tan  -400.000000 =>  0000000000000 00
Grad: tan  -350.000000 =>  1000000000000 00
Grad: tan  -300.000000 =>  9999999999999 99
Grad: tan  -250.000000 => -9999999999997-01
Grad: tan  -200.000000 =>  0000000000000 00
Grad: tan  -150.000000 =>  1000000000001 00
Grad: tan  -100.000000 =>  9999999999999 99
Grad: tan   -50.000000 => -9999999999993-01
Grad: tan     0.000000 =>  0000000000000 00
Grad: tan    50.000000 =>  1000000000000 00
Grad: tan   100.000000 =>  9999999999999 99
Grad: tan   150.000000 => -9999999999997-01
Grad: tan   200.000000 =>  0000000000000 00
Grad: tan   250.000000 =>  1000000000001 00
Grad: tan   300.000000 =>  9999999999999 99
Grad: tan   350.000000 => -9999999999993-01
Grad: tan   400.000000 =>  0000000000000 00
Grad: tan   450.000000 =>  1000000000000 00
Grad: tan   500.000000 =>  9999999999999 99
Grad: tan   550.000000 => -9999999999997-01
Grad: tan   600.000000 =>  0000000000000 00
Grad: tan   650.000000 =>  1000000000001 00
Grad: tan   700.000000 =>  9999999999999 99
Grad: tan   750.000000 => -9999999999993-01

Rad:  sin   -12.566371 =>  1776356839400-15
Rad:  sin   -11.780972 =>  7071067811884-01
Rad:  sin   -10.995574 =>  1000000000000 00
Rad:  sin   -10.210176 =>  7071067811882-01
Rad:  sin    -9.424778 => -2077112855299-13
Rad:  sin    -8.639380 => -7071067811864-01
Rad:  sin    -7.853982 => -1000000000000 00
Rad:  sin    -7.068583 => -7071067811866-01
Rad:  sin    -6.283185 =>  0000000000000 00
Rad:  sin    -5.497787 =>  7071067811862-01
Rad:  sin    -4.712389 =>  1000000000000 00
Rad:  sin    -3.926991 =>  7071067811868-01
Rad:  sin    -3.141593 =>  0000000000000 00
Rad:  sin    -2.356194 => -7071067811864-01
Rad:  sin    -1.570796 => -1000000000000 00
Rad:  sin    -0.785398 => -7071067811863-01
Rad:  sin     0.000000 =>  0000000000000 00
Rad:  sin     0.785398 =>  7071067811866-01
Rad:  sin     1.570796 =>  1000000000000 00
Rad:  sin     2.356194 =>  7071067811861-01
Rad:  sin     3.141593 =>  0000000000000 00
Rad:  sin     3.926991 => -7071067811871-01
Rad:  sin     4.712389 => -1000000000000 00
Rad:  sin     5.497787 => -7071067811859-01
Rad:  sin     6.283185 =>  0000000000000 00
Rad:  sin     7.068583 =>  7071067811869-01
Rad:  sin     7.853982 =>  1000000000000 00
Rad:  sin     8.639380 =>  7071067811861-01
Rad:  sin     9.424778 => -2059349286905-13
Rad:  sin    10.210176 => -7071067811885-01
Rad:  sin    10.995574 => -1000000000000 00
Rad:  sin    11.780972 => -7071067811881-01

Rad:  cos   -12.566371 =>  1000000000000 00
Rad:  cos   -11.780972 =>  7071067811847-01
Rad:  cos   -10.995574 =>  4895700681727-12
Rad:  cos   -10.210176 => -7071067811849-01
Rad:  cos    -9.424778 => -1000000000000 00
Rad:  cos    -8.639380 => -7071067811867-01
Rad:  cos    -7.853982 =>  0000000000000 00
Rad:  cos    -7.068583 =>  7071067811865-01
Rad:  cos    -6.283185 =>  1000000000000 00
Rad:  cos    -5.497787 =>  7071067811869-01
Rad:  cos    -4.712389 => -1038556427650-13
Rad:  cos    -3.926991 => -7071067811863-01
Rad:  cos    -3.141593 => -1000000000000 00
Rad:  cos    -2.356194 => -7071067811867-01
Rad:  cos    -1.570796 =>  0000000000000 00
Rad:  cos    -0.785398 =>  7071067811868-01
Rad:  cos     0.000000 =>  1000000000000 00
Rad:  cos     0.785398 =>  7071067811865-01
Rad:  cos     1.570796 =>  0000000000000 00
Rad:  cos     2.356194 => -7071067811870-01
Rad:  cos     3.141593 => -1000000000000 00
Rad:  cos     3.926991 => -7071067811860-01
Rad:  cos     4.712389 =>  3097905714555-13
Rad:  cos     5.497787 =>  7071067811872-01
Rad:  cos     6.283185 =>  1000000000000 00
Rad:  cos     7.068583 =>  7071067811862-01
Rad:  cos     7.853982 => -1029674643453-13
Rad:  cos     8.639380 => -7071067811870-01
Rad:  cos     9.424778 => -1000000000000 00
Rad:  cos    10.210176 => -7071067811846-01
Rad:  cos    10.995574 =>  5309346895947-12
Rad:  cos    11.780972 =>  7071067811850-01

Rad:  tan   -12.566371 =>  1776356839400-15
Rad:  tan   -11.780972 =>  1000000000005 00
Rad:  tan   -10.995574 =>  2042608535551 11
Rad:  tan   -10.210176 => -1000000000005 00
Rad:  tan    -9.424778 =>  2077112855299-13
Rad:  tan    -8.639380 =>  9999999999995-01
Rad:  tan    -7.853982 =>  9999999999999 99
Rad:  tan    -7.068583 => -1000000000000 00
Rad:  tan    -6.283185 =>  0000000000000 00
Rad:  tan    -5.497787 =>  9999999999991-01
Rad:  tan    -4.712389 => -9628749804796 12
Rad:  tan    -3.926991 => -1000000000001 00
Rad:  tan    -3.141593 =>  0000000000000 00
Rad:  tan    -2.356194 =>  9999999999995-01
Rad:  tan    -1.570796 =>  9999999999999 99
Rad:  tan    -0.785398 => -9999999999993-01
Rad:  tan     0.000000 =>  0000000000000 00
Rad:  tan     0.785398 =>  1000000000000 00
Rad:  tan     1.570796 =>  9999999999999 99
Rad:  tan     2.356194 => -9999999999987-01
Rad:  tan     3.141593 =>  0000000000000 00
Rad:  tan     3.926991 =>  1000000000002 00
Rad:  tan     4.712389 => -3227987202134 12
Rad:  tan     5.497787 => -9999999999983-01
Rad:  tan     6.283185 =>  0000000000000 00
Rad:  tan     7.068583 =>  1000000000001 00
Rad:  tan     7.853982 => -9711805630630 12
Rad:  tan     8.639380 => -9999999999987-01
Rad:  tan     9.424778 =>  2059349286905-13
Rad:  tan    10.210176 =>  1000000000006 00
Rad:  tan    10.995574 => -1883470829083 11
Rad:  tan    11.780972 => -1000000000004 00

